gprofiler/profilers/perf.py

```diff
@@ -81,39 +85,66 @@ def _get_perf_cmd(self) -> List[str]:
             str(self._mmap_sizes[self._type]),
         ] + self._extra_args
 
-    def check_if_needs_restart(self) -> None:
-        """Checks if perf used memory exceeds threshold, and if it does, restarts perf"""
-        assert self._process is not None
-        if (
-            time.monotonic() - self._start_time >= self._restart_after_s
-            and Process(self._process.pid).memory_info().rss >= self._perf_memory_usage_threshold
-        ):
-            self.stop()
-            self.start()
-
     def start(self) -> None:
-        logger.info(f"Starting perf ({self._type} mode)")
+        logger.info(f"Starting {self._log_name}")
+        # remove old files, should they exist from previous runs
+        remove_path(self._output_path, missing_ok=True)
         process = start_process(self._get_perf_cmd(), via_staticx=False)
         try:
             wait_event(self._poll_timeout_s, self._stop_event, lambda: os.path.exists(self._output_path))
             self.start_time = time.monotonic()
         except TimeoutError:
             process.kill()
             assert process.stdout is not None and process.stderr is not None
-            logger.critical(f"perf failed to start. stdout {process.stdout.read()!r} stderr {process.stderr.read()!r}")
+            logger.critical(
+                f"{self._log_name} failed to start", stdout=process.stdout.read(), stderr=process.stderr.read()
+            )
             raise
         else:
             self._process = process
             os.set_blocking(self._process.stdout.fileno(), False)  # type: ignore
             os.set_blocking(self._process.stderr.fileno(), False)  # type: ignore
-            logger.info(f"Started perf ({self._type} mode)")
+            logger.info(f"Started {self._log_name}")
 
     def stop(self) -> None:
         if self._process is not None:
             self._process.terminate()  # okay to call even if process is already dead
-            self._process.wait()
+            exit_code = self._process.wait()
             self._process = None
-            logger.info(f"Stopped perf ({self._type} mode)")
+            logger.info(f"Stopped {self._log_name}", exit_code=exit_code)
+
+    def is_running(self) -> bool:
+        """
+        Is perf running? returns False if perf is stopped OR if process exited since last check
+        """
+        return self._process is not None and self._process.poll() is None
+
+    def restart(self) -> None:
+        self.stop()
+        self.start()
+
+    def restart_if_not_running(self) -> None:
+        """
+        Restarts perf if it was stopped for whatever reason.
+        """
+        if not self.is_running():
+            logger.warning(f"{self._log_name} not running (unexpectedly), restarting...")
+            self.restart()
+
+    def check_if_needs_memory_restart(self) -> None:
```

Comment:
This doesn't just check, it restarts. So `restart_if_rss_exceeded`?

Reply:
Hm. Yeah makes sense